### PR TITLE
minirouter: Added new features to minirouter routing protocol and interfaces

### PR DESCRIPTION
**Pull request by aherna**
Added Loopback,routerid,custom ospf advertisements, named static routes, and base BGP

### CODE REVIEW

File: src/minimega/router.go [hunk 1]
```diff
@@ -46,6 +52,16 @@ type dhcp struct {
 	static map[string]string
 }
 
+type bgp struct {
+	bgpprocessname string
```
Comment (jcrussell) [inline src/minimega/router.go]: `bgpprocessname` in `bgp` struct should simply be `processName`.

File: src/minimega/router.go [hunk 2]
```diff
@@ -46,6 +52,16 @@ type dhcp struct {
...
+	localip        string
```
Comment (jcrussell) [inline src/minimega/router.go]: Go convention is [MixedCaps](https://golang.org/doc/effective_go.html#mixed-caps).

File: src/minimega/router.go [hunk 3]
```diff
@@ -385,13 +507,59 @@ func (r *Router) InterfaceDel(n string, i string) error {
...
+			}
+		}
+	}
+	return false
+}
+
+//Generates router ID based on established rules
+//Manually configure has highest preference -> Highest loopback -> Highest interface
+func (r *Router) setRouterID() {
+	if strings.Compare(r.routerid, "0.0.0.0") == 0 {
```
Comment (jcrussell) [inline src/minimega/router.go]: Return if `!= 0` to remove level.

File: src/minimega/router.go [hunk 4]
```diff
@@ -385,13 +507,59 @@ func (r *Router) InterfaceDel(n string, i string) error {
...
+		if len(r.Loopbacks) > 0 {
+			for _, ip := range r.Loopbacks {
+				log.Debug("compare loopback ip %v with current router id %v", ip, r.routerid)
+				if bytes.Compare(net.ParseIP(strings.Split(ip, "/")[0]), net.ParseIP(r.routerid)) == 1 {
+					log.Debug("found new router id %v", ip)
+					r.routerid = strings.Split(ip, "/")[0]
+				}
+			}
+		} else {
```
Comment (jcrussell) [inline src/minimega/router.go]: Return before else to remove level.

File: src/minimega/router_test.go [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+	}
+}
+
+func configmatch(test string, t int) (bool, string) {
+	var control map[string]bool
+	testarr := strings.Split(test, "\n")
+	switch t {
+	case 0:
+		control = map[string]bool{
+			"log level error":          true,
```
Comment (jcrussell) [inline src/minimega/router_test.go]: `gofmt` this.

File: src/minirouter/bird.go [hunk 6]
```diff
@@ -32,89 +36,175 @@ var (
...
-	Area       string
-	Interfaces map[string]bool // bool placeholder for later options
+	Area           string
+	Interfaces     map[string]bool // bool placeholder for later options
+	Prefixes       map[string]bool
+	Filternetworks map[string]bool
+}
+
+type BGP struct {
+	Bgpprocessname    string
```
Comment (jcrussell) [inline src/minirouter/bird.go]: See above.

File: src/minirouter/bird.go [hunk 7]
```diff
@@ -32,89 +36,175 @@ var (
...
 		}
-
-		o := OSPFFindOrCreate(area)
-		o.Interfaces[iface] = true
+	} else if c.BoolArgs["bgp"] {
+		var ip string
+		processname := c.StringArgs["processname"]
+		log.Debugln("bird: Looking for Bgp process")
+		b := bgpFindOrCreate(processname)
+		log.Debug("bird: Found BGP process %", b.Bgpprocessname)
```
Comment (jcrussell) [inline src/minirouter/bird.go]: Missing format arg.

File: src/minirouter/bird.go [hunk 8]
```diff
@@ -32,89 +36,175 @@ var (
...
+				b.Localas = as
+			} else if c.BoolArgs["neighbor"] {
+				log.Debug("bird: Setting neighbor IP %v and AS %v\n", ip, as)
+				b.Neighborip = ip
+				b.Neighboras = as
+			}
+		} else if c.BoolArgs["rrclient"] {
+			b.Routereflector = true
+		} else if c.BoolArgs["filter"] {
+			log.Debug("bird: adding filter %", c.StringArgs["filtername"])
```
Comment (jcrussell) [inline src/minirouter/bird.go]: Missing format arg.

File: src/minimega/router.go (hunk 2 above)
Comment (aherna) [reply src/minimega/router.go]: resolved in latest commit

File: src/minimega/router.go (hunk 4 above)
Comment (aherna) [reply src/minimega/router.go]: resolved in latest commit

File: src/minimega/router_test.go (hunk 5 above)
Comment (aherna) [reply src/minimega/router_test.go]: resolved in latest commit

File: src/minirouter/bird.go (hunk 6 above)
Comment (aherna) [reply src/minirouter/bird.go]: resolved in latest commit

File: src/minirouter/bird.go (hunk 8 above)
Comment (aherna) [reply src/minirouter/bird.go]: resolved in latest commit

File: src/minirouter/bird.go (hunk 7 above)
Comment (aherna) [reply src/minirouter/bird.go]: resolved in latest commit

File: src/minimega/router_cli.go [hunk 9]
```diff
@@ -68,8 +69,9 @@ router takes a number of subcommands:
 		Patterns: []string{
```
Comment (jcrussell) [inline src/minimega/router_cli.go]: Need to add documentation for new subcommands.

File: src/minimega/router_cli.go [hunk 10]
```diff
@@ -232,12 +277,20 @@ func cliClearRouter(ns *Namespace, c *minicli.Command, resp *minicli.Response) e
 	}
 
 	if c.BoolArgs["interface"] {
-		network := c.StringArgs["network"]
-		ip := c.StringArgs["IPv4/MASK"]
-
-		err := rtr.InterfaceDel(network, ip)
-		if err != nil {
-			return err
+		if c.BoolArgs["lo"] {
```
Comment (jcrussell) [inline src/minimega/router_cli.go]: The if/else block are the same except for passing true/false to `InterfaceDel`.

File: src/minirouter/ip.go [hunk 11]
```diff
@@ -109,72 +110,64 @@ func handleIP(c *minicli.Command, r chan<- minicli.Responses) {
...
-	// get an interface from the index
-	eth, err := findEth(idx)
-	if err != nil {
-		return err
-	}
 
-	// TODO: what's the right way to remove a dhcp interface?
+	// TODO: Need to send release ip reservation to dhcp server
 	if ip == "dhcp" {
+		out, err := exec.Command("dhclient -r", iface).CombinedOutput()
```
Comment (jcrussell) [inline src/minirouter/ip.go]: This should be `exec.Command("dhclient", "-r", iface).CombinedOutput()`

File: doc/content/articles/router.article [hunk 12]
```diff
@@ -89,6 +89,10 @@ Multiple addresses can be added to the same interface as well:
 	minimega$ router foo interface 0 10.0.0.1/24
 	minimega$ router foo interface 0 2001:1::1/64
 
+Minirouter can also assign loopback interfaces. These types of interfaces are useful in routing processes (ie OSPF)
```
Comment (jcrussell) [inline doc/content/articles/router.article]: Lower case for `minirouter`.

File: doc/content/articles/router.article [hunk 13]
```diff
@@ -142,7 +146,7 @@ the record:
 
 `minirouter` uses the [[http://bird.network.cz/][bird routing daemon]] to
 provide routing using a variety of protocols. Currently, `minirouter` only
-supports static and OSPF routes. Support for BGP and others are planned.
+supports static,OSPF and bgp routes.
```
Comment (jcrussell) [inline doc/content/articles/router.article]: Missing space between static and OSPF.

File: doc/content/articles/router.article [hunk 14]
```diff
@@ -163,13 +167,18 @@ IPv6 routes are added in the same way:
 
 	minimega$ router foo route static 2001:1:2:3::/64 2001:1:2:3::1
 
+you can also give names to static routes so you can combine multple routes under a single name. By doing so you can create a sort of filter to be used in routing processes like OSPF and BGP. For example setting a static route with the name bar-route
```
Comment (jcrussell) [inline doc/content/articles/router.article]: Capitalize. For example needs a comma after.

File: doc/content/articles/router.article [hunk 15]
```diff
@@ -163,13 +167,18 @@ IPv6 routes are added in the same way:
...
+
 ** OSPF
 
 `minirouter` provides basic support for OSPF and OSPFv3 (IPv6 enabled OSPF) by
 specifying the OSPF area and *interface* to include in the area. OSPF generally
 supports specifying networks and many other options, which `minirouter` may add
-in the future. For now, specifying an interface (and all of the networks on
+in the future. Specifying an interface (and all of the networks on
 that interface) is provided. Both OSPF and OSPFv3 are enabled by `minirouter`.
+You can also specify specific networks or static routes into the OSPF process
```
Comment (jcrussell) [inline doc/content/articles/router.article]: Missing .

File: doc/content/articles/router.article (hunk 12 above)
Comment (jcrussell) [reply doc/content/articles/router.article]: ie is an abbreviation -- should be i.e.

File: doc/content/articles/router.article [hunk 16]
```diff
@@ -183,3 +192,43 @@ router VM to area 0 in an OSPF route:
 	# add interface 'c', index 2
 	minimega$ router foo route ospf 0 2
 
+Now lets say you want to advertise a network that is on index 1 but you do not want interface 1 to participate in OSPF (ie its an Internet facing interface)
```
Comment (jcrussell) [inline doc/content/articles/router.article]: Another ie.

File: doc/content/articles/router.article [hunk 17]
```diff
@@ -183,3 +192,43 @@ router VM to area 0 in an OSPF route:
...
+	
+	# advertise network 10.0.0.0/24 in OSPF area 0
+	minimega$router foo route ospf 0 export 10.0.0.0/24
```
Comment (jcrussell) [inline doc/content/articles/router.article]: Add space between `minimega$` and `router`.

File: doc/content/articles/router.article [hunk 18]
```diff
@@ -183,3 +192,43 @@ router VM to area 0 in an OSPF route:
...
 
+Now lets say you want to advertise a network that is on index 1 but you do not want interface 1 to participate in OSPF (ie its an Internet facing interface)
+	
+	# advertise network 10.0.0.0/24 in OSPF area 0
+	minimega$router foo route ospf 0 export 10.0.0.0/24 
+
+Say you want to advertise a specific static route(s) into the ospf process. The export command provides this functionality
+
+	# advertise network default route 0.0.0.0/0 in OSPF area 0
+	minimega$router foo route static 0.0.0.0/0 1.2.3.254 default-route
```
Comment (jcrussell) [inline doc/content/articles/router.article]: Add space here too.

File: doc/content/articles/router.article [hunk 19]
```diff
@@ -183,3 +192,43 @@ router VM to area 0 in an OSPF route:
...
+	minimega$router foo route ospf export default-route
+
+** BGP
+
+‘minirouter’ provides basic support for BGP on IPv4 Networks. Each BGP peering relationship or process needs to have a name, local information, neighbor information and whats being exported. The local and neighbor information will be IP address and AS numbers. You can additionally set up whats being exported ie all learned routes or just routes specified. The filter keyword allows you to use named static routes for simple route filtering. More advanced route filtering is planned. Route reflection can also be configured if you want to configure a particular router to be a route reflector server by default all routers are BGP route clients.
```
Comment (jcrussell) [inline doc/content/articles/router.article]: Drop the quotes around minirouter.

File: doc/content/articles/router.article [hunk 20]
```diff
@@ -183,3 +192,43 @@ router VM to area 0 in an OSPF route:
...
+
+For example R1 has an ip of 10.0.0.1 in AS 100 and R2 has an IP of 20.0.0.1 in AS 20. We would like R1 just to announce the 10.0.0.0/24 network to R2 while on R2 we want to advertise all learned routes. We could also optionally add additional routes by adding more export statements to R2 
+
+To configure BGP:
+	
+	#Set up routes to be advertised by BGP
+	minimega$router R1 route static 10.0.0.0/24 0 bar-route
```
Comment (jcrussell) [inline doc/content/articles/router.article]: Add space.

File: doc/content/articles/router.article [hunk 21]
```diff
@@ -183,3 +192,43 @@ router VM to area 0 in an OSPF route:
...
+
+	# advertise network default route 0.0.0.0/0 in OSPF area 0
+	minimega$router foo route static 0.0.0.0/0 1.2.3.254 default-route
+	minimega$router foo route ospf export default-route
+
+** BGP
+
+‘minirouter’ provides basic support for BGP on IPv4 Networks. Each BGP peering relationship or process needs to have a name, local information, neighbor information and whats being exported. The local and neighbor information will be IP address and AS numbers. You can additionally set up whats being exported ie all learned routes or just routes specified. The filter keyword allows you to use named static routes for simple route filtering. More advanced route filtering is planned. Route reflection can also be configured if you want to configure a particular router to be a route reflector server by default all routers are BGP route clients.
+
+For example R1 has an ip of 10.0.0.1 in AS 100 and R2 has an IP of 20.0.0.1 in AS 20. We would like R1 just to announce the 10.0.0.0/24 network to R2 while on R2 we want to advertise all learned routes. We could also optionally add additional routes by adding more export statements to R2
```
Comment (jcrussell) [inline doc/content/articles/router.article]: For example,

File: src/minimega/router_cli.go [hunk 22]
```diff
@@ -30,12 +31,14 @@ router takes a number of subcommands:
 
 - 'interface': Set IPv4 or IPv6 addresses, or configure an interface to assign
   using DHCP. The interface field is an integer index of the interface defined
-  with 'vm config net'. For example, to configure the second interface of the
-  router with a static IP:
+  with 'vm config net'. you could also specify if that interface will be a
```
Comment (jcrussell) [inline src/minimega/router_cli.go]: Capitalize you.

File: src/minimega/router_cli.go [hunk 23]
```diff
@@ -55,21 +58,57 @@ router takes a number of subcommands:
...
-  interface described in 'vm config net'. For example, to enable OSPF on area 0
-  for both interfaces of a router:
+- 'route': Set static, OSPF, or BGP routes. Static routes include a subnet,
+  next-hop, and optionally a name for this router. For example to specify a 
+  static route(s):
+
+    router foo route static 0.0.0.0/0 10.0.0.1 default-route
+  
+  OSPF routes include an area and a network index corresponding to the interface described in 
+  'vm config net'. You can also specify what networks to advertise using the export command
```
Comment (jcrussell) [inline src/minimega/router_cli.go]: Missing .

File: src/minimega/router_cli.go [hunk 24]
```diff
@@ -55,21 +58,57 @@ router takes a number of subcommands:
...
+  
+  For example, to enable OSPF on area 0 for both interfaces of a router:
 
 	vm config net 100 200
 	# ...
 	router foo route ospf 0 0
 	router foo route ospf 0 1
+
+  For example to advertrise specific networks, advertise a static route or use a static route as a filter:
```
Comment (jcrussell) [inline src/minimega/router_cli.go]: For example,

File: src/minimega/router_cli.go [hunk 25]
```diff
@@ -55,21 +58,57 @@ router takes a number of subcommands:
...
+	
+    router foo route static 11.0.0.0/24 0 bar-route
```
Comment (jcrussell) [inline src/minimega/router_cli.go]: Indentation off.

File: src/minimega/router_cli.go [hunk 26]
```diff
@@ -55,21 +58,57 @@ router takes a number of subcommands:
...
+	router foo route static 12.0.0.0/24 0 bar-route
+	router foo route ospf 0 export 10.0.0.0/24
+	router foo route ospf 0 export default-route
+	router foo route ospf 0 export bar-route
+
+  To configure BGP must specify the process name for the specific bgp context, local ip address and AS,
+  Neighbor ip address and AS, and what networks need to be advertised
+
+  For example local router is in AS 100 with an ip 10.0.0.1 and bgp peer is in AS 200 with an ip of 20.0.0.1
```
Comment (jcrussell) [inline src/minimega/router_cli.go]: For example,

File: src/minirouter/ip.go (hunk 11 above)
Comment (aherna) [reply src/minirouter/ip.go]: Done

**Comment by jcrussell** [comment]
@aherna: anything else that you want to do on this before merging?

**Comment by aherna** [comment]
> @aherna: anything else that you want to do on this before merging?

Good on my end

**Pull request merged**